OpenRA.Mods.Common/UpdateRules/Rules/20221203/AbstractDocking.cs

```diff
@@ -0,0 +1,163 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class AbstractDocking : UpdateRule
+	{
+		readonly string[] refineyValues = { "DockAngle", "IsDragRequired", "DragOffset", "DragLength" };
+		readonly string[] harvesterValues = { "EnterCursor", "EnterBlockedCursor" };
+
+		readonly Dictionary<string, List<MiniYamlNode>> refineryNodes = new Dictionary<string, List<MiniYamlNode>>();
+		public override string Name => "Docking was abstracted from Refinery & Harvester.";
+
+		public override string Description =>
+			"Fields moved from Refinery to new trait DockHost, fields moved from Harvester to new trait DockClientManager and to DockHost";
+
+		public override IEnumerable<string> BeforeUpdateActors(ModData modData, List<MiniYamlNode> resolvedActors)
+		{
+			var harvesters = new Dictionary<string, HashSet<string>>();
+			var refineries = new List<string>();
+			foreach (var actorNode in resolvedActors)
+			{
+				var harvesterNode = actorNode.ChildrenMatching("Harvester", includeRemovals: false).FirstOrDefault();
+				if (harvesterNode != null)
+					harvesters[actorNode.Key] = harvesterNode.ChildrenMatching("DeliveryBuildings", includeRemovals: false).FirstOrDefault()?.NodeValue<HashSet<string>>() ?? new HashSet<string>();
+
+				var refineryNode = actorNode.ChildrenMatching("Refinery", includeRemovals: false).FirstOrDefault() ??
+					actorNode.ChildrenMatching("TiberianSunRefinery", includeRemovals: false).FirstOrDefault();
+				if (refineryNode != null)
+					refineries.Add(actorNode.Key);
+			}
+
+			foreach (var harvester in harvesters)
+			{
+				foreach (var deliveryBuilding in harvester.Value)
+				{
+					foreach (var refinery in refineries)
+					{
+						if (refinery == deliveryBuilding)
+						{
+							if (!refineryNodes.ContainsKey(refinery))
+								refineryNodes[refinery] = new List<MiniYamlNode>();
+
+							var node = new MiniYamlNode("DockType", deliveryBuilding.ToString());
+							if (!refineryNodes[refinery].Any(n => n.Key == node.Key))
```

Comment:
This is not checking for duplicates

Reply:
Sorry I meant to say *permit* duplicates by only adding entries where they don't already exist. However now that I see these are UpdateRules it probably doesn't matter, since I presume it is only here for legacy compatibility reasons.